### ISSUE TITLE
使用nrtr训练文本识别，验证准确率过低的问题？

### ISSUE THREAD

**Issue by guanrenxue** [PaddlePaddle/PaddleOCR#4769]
在使用rec_mtb_nrtr.yml配置文件进行文本识别训练时，经过50次的迭代训练，训练过程中的准确率已经达到0.8左右，为什么在使用验证集进行评价时，得到的acc分数为0.02左右，如下是训练情况：
val model:: 100%|██████████| 39/39 [3:06:17<00:00, 275.56s/it]
eval model:: 100%|██████████| 39/39 [3:06:18<00:00, 286.62s/it]
[2021/11/26 02:56:08] root INFO: cur metric, acc: 0.024483243026289386, norm_edit_dis: 0.45585130064280965, fps: 0.44601847659123284
[2021/11/26 02:56:11] root INFO: save best model is to ./output/rec_nrtr/nrtr/best_accuracy
[2021/11/26 02:56:11] root INFO: best metric, acc: 0.024483243026289386, norm_edit_dis: 0.45585130064280965, fps: 0.44601847659123284, best_epoch: 52
[2021/11/26 02:56:24] root INFO: epoch: [52/200], iter: 39790, lr: 0.000430, loss: 0.797760, acc: 0.867188, norm_edit_dis: 0.993823, reader_cost: 0.00059 s, batch_cost: 1.22175 s, samples: 1280, ips: 104.76796
[2021/11/26 02:56:37] root INFO: epoch: [52/200], iter: 39800, lr: 0.000429, loss: 0.798407, acc: 0.867188, norm_edit_dis: 0.994110, reader_cost: 0.00042 s, batch_cost: 1.22400 s, samples: 1280, ips: 104.57533
.........
请问遇到这种情况怎么解决？

**Comment by Topdu** [comment]
请问您用的什么数据集呢？数据集的样本量有多少？字典是否和数据集的label对应？

如果数据集是英文，数据集label对应正确，样本量足够多，这个模型很明显是过拟合了，但是过拟合的非常严重，不符合经验，感觉是数据集的错误，可以打印一下加载后的label，和模型预测的label，确保加载的数据和label是没有错误的。

有问题随时交流

**Comment by guanrenxue** [comment]
> 
如下是我的配置文件：
Global:
  use_gpu: True
  epoch_num: 200
  log_smooth_window: 20
  print_batch_step: 10
  save_model_dir: ./output/rec_nrtr/nrtr/
  save_epoch_step: 3
  # evaluation is run every 2000 iterations
  eval_batch_step: [11700, 1560]
  cal_metric_during_train: True
  pretrained_model:
  checkpoints: 
  save_inference_dir:
  use_visualdl: True
  infer_img: doc/imgs_words_en/word_10.png
  # for data or label process
  character_dict_path: ./train_data/rec_dataset/ZW/alphabet0830.txt
  character_type: ch  
  max_text_length: 118
  infer_mode: False
  use_space_char: True
  save_res_path: ./output/rec_nrtr/predicts_nrtr.txt

Optimizer:
  name: Adam
  beta1: 0.9
  beta2: 0.99
  clip_norm: 5.0
  lr:
    name: Cosine
    learning_rate: 0.0005
    warmup_epoch: 2
  regularizer:
    name: 'L2'
    factor: 0.

Architecture:
  model_type: rec
  algorithm: NRTR
  in_channels: 1
  Transform:
  Backbone:
    name: MTB
    cnn_num: 2
  Head:
    name: Transformer
    d_model: 512
    num_encoder_layers: 6
    beam_size: 10 # When Beam size is greater than 0, it means to use beam search when evaluation.
    

Loss:
  name: NRTRLoss
  smoothing: True

PostProcess:
  name: NRTRLabelDecode

Metric:
  name: RecMetric
  main_indicator: acc

Train:
  dataset:
    name: SimpleDataSet
    data_dir: ./train_data/rec_dataset/ZW/
    label_file_list: ["./train_data/rec_dataset/ZW/train_0830.txt"]
    transforms:
      - NRTRDecodeImage: # load image
          img_mode: BGR
          channel_first: False
      - NRTRLabelEncode: # Class handling label
      - NRTRRecResizeImg:
          image_shape: [640, 32]  #[100, 32]
          resize_type: PIL # PIL or OpenCV
      - KeepKeys:
          keep_keys: ['image', 'label', 'length'] # dataloader will return list in this order
  loader:
    shuffle: True
    batch_size_per_card: 128
    drop_last: True
    num_workers: 16
    use_shared_memory: False

Eval:
  dataset:
    name: SimpleDataSet
    data_dir: ./train_data/rec_dataset/ZW/
    label_file_list: ["./train_data/rec_dataset/ZW/val_0830.txt"]
    transforms:
      - NRTRDecodeImage: # load image
          img_mode: BGR
          channel_first: False
      - NRTRLabelEncode: # Class handling label
      - NRTRRecResizeImg:
          image_shape: [640, 32]   #[100, 32]
          resize_type: PIL # PIL or OpenCV
      - KeepKeys:
          keep_keys: ['image', 'label', 'length'] # dataloader will return list in this order
  loader:
    shuffle: False
    drop_last: False
    batch_size_per_card: 128
    num_workers: 16
    use_shared_memory: False

数据采用的是英文数据集，训练样本30万左右，验证集5千，配置文件中训练集地址和验证集地址是没有问题的，之前使用CRNN跑通过，我重新修改控制模型的dim和层数，重新训练，看看是不是过拟合了。
此外在使用验证集进行评价时，花费时间巨长，大概需要3个小时，这个怎么解决？

**Comment by Topdu** [comment]
配置没问题，建议训练的时候关闭beam搜索，训练结束再用beam搜索，beam搜非常的耗时间；再看一下max_text_length: 118是否被使用，因为label加载的时候默认最长25个字符！---- 回复的原邮件 ***@***.***>发送日期2021年11月29日 09:44 ***@***.***> ***@***.******@***.***>主题Re: [PaddlePaddle/PaddleOCR] 使用nrtr训练文本识别，验证准确率过低的问题？ (Issue #4769)





如下是我的配置文件：

Global:

use_gpu: True

epoch_num: 200

log_smooth_window: 20

print_batch_step: 10

save_model_dir: ./output/rec_nrtr/nrtr/

save_epoch_step: 3

evaluation is run every 2000 iterations

eval_batch_step: [11700, 1560]

cal_metric_during_train: True

pretrained_model:

checkpoints:

save_inference_dir:

use_visualdl: True

infer_img: doc/imgs_words_en/word_10.png

for data or label process

character_dict_path: ./train_data/rec_dataset/ZW/alphabet0830.txt

character_type: ch

max_text_length: 118

infer_mode: False

use_space_char: True

save_res_path: ./output/rec_nrtr/predicts_nrtr.txt

Optimizer:

name: Adam

beta1: 0.9

beta2: 0.99

clip_norm: 5.0

lr:

name: Cosine

learning_rate: 0.0005

warmup_epoch: 2

regularizer:

name: 'L2'

factor: 0.

Architecture:

model_type: rec

algorithm: NRTR

in_channels: 1

Transform:

Backbone:

name: MTB

cnn_num: 2

Head:

name: Transformer

d_model: 512

num_encoder_layers: 6

beam_size: 10 # When Beam size is greater than 0, it means to use beam search when evaluation.

Loss:

name: NRTRLoss

smoothing: True

PostProcess:

name: NRTRLabelDecode

Metric:

name: RecMetric

main_indicator: acc

Train:

dataset:

name: SimpleDataSet

data_dir: ./train_data/rec_dataset/ZW/

label_file_list: ["./train_data/rec_dataset/ZW/train_0830.txt"]

transforms:

- NRTRDecodeImage: # load image

img_mode: BGR

channel_first: False

- NRTRLabelEncode: # Class handling label

- NRTRRecResizeImg:

image_shape: [640, 32]  #[100, 32]

resize_type: PIL # PIL or OpenCV

- KeepKeys:

keep_keys: ['image', 'label', 'length'] # dataloader will return list in this order

loader:

shuffle: True

batch_size_per_card: 128

drop_last: True

num_workers: 16

use_shared_memory: False

Eval:

dataset:

name: SimpleDataSet

data_dir: ./train_data/rec_dataset/ZW/

label_file_list: ["./train_data/rec_dataset/ZW/val_0830.txt"]

transforms:

- NRTRDecodeImage: # load image

img_mode: BGR

channel_first: False

- NRTRLabelEncode: # Class handling label

- NRTRRecResizeImg:

image_shape: [640, 32]   #[100, 32]

resize_type: PIL # PIL or OpenCV

- KeepKeys:

keep_keys: ['image', 'label', 'length'] # dataloader will return list in this order

loader:

shuffle: False

drop_last: False

batch_size_per_card: 128

num_workers: 16

use_shared_memory: False

数据采用的是英文数据集，训练样本30万左右，验证集5千，配置文件中训练集地址和验证集地址是没有问题的，之前使用CRNN跑通过，我重新修改控制模型的dim和层数，重新训练，看看是不是过拟合了。

此外在使用验证集进行评价时，花费时间巨长，大概需要3个小时，这个怎么解决？



—You are receiving this because you commented.Reply to this email directly, view it on GitHub, or unsubscribe.Triage notifications on the go with GitHub Mobile for iOS or Android.



[

{

***@***.***": "http://schema.org",

***@***.***": "EmailMessage",

"potentialAction": {

***@***.***": "ViewAction",

"target": "https://github.com/PaddlePaddle/PaddleOCR/issues/4769#issuecomment-981220505",

"url": "https://github.com/PaddlePaddle/PaddleOCR/issues/4769#issuecomment-981220505",

"name": "View Issue"

},

"description": "View this Issue on GitHub",

"publisher": {

***@***.***": "Organization",

"name": "GitHub",

"url": "https://github.com"

}

}

]
